Add no-server-header support for Web Socket implementations

iudeen opened the pull request.

iudeen committed:
added logic to accept default headers in websocket

iudeen committed:
added default_headers to class init

Commit:
fix

Commit:
docs: added limitation about Websockets

Commit:
docs: added limitation about Websockets

iudeen committed:
Update docs/settings.md

adopted change suggested.

Co-authored-by: Marcelo Trylesinski <marcelotryle@gmail.com>

Commit:
docs: moved websocket limitation

as per suggestion, moved the limitation to --no-server-header area

Kludex committed:
Update docs/settings.md

Commit:
Merge remote-tracking branch 'origin/master' into patch/websocket-server-name

Commit:
feat(websockets): added server header support for websockets

Added logic to add server header support for websockets and also to pass default headers

iudeen committed:
Merge branch 'encode:master' into master

Commit:
feat(websockets-wsproto): added server header support for websockets

Added logic to add server header support for websockets and also to pass default headers

iudeen committed:
added logic to accept default headers in websocket (#1)

* added logic to accept default headers in websocket

* added default_headers to class init

* fix

* feat(websockets): added server header support for websockets

Added logic to add server header support for websockets and also to pass default headers

* feat(websockets-wsproto): added server header support for websockets

Added logic to add server header support for websockets and also to pass default headers

Co-authored-by: Irfanuddin <irfanuddin@knowledgelens.com>

Commit:
style(websockets): formatting

Commit:
Merge branch 'patch/websocket-server-name'

# Conflicts:
#	uvicorn/protocols/websockets/websockets_impl.py
#	uvicorn/protocols/websockets/wsproto_impl.py

Commit:
style(websockets): formatting

Commit:
style(websockets): formatting

iudeen commented:
I am not sure if these features are needed at this point of time, but I went ahead and added as I was not able to conclusively find a reason why server headers were not added to Websocket protocols.

Note: Also, we can keep this in WIP till Websocket library provides an API to disable server header completely.

Kludex reviewed (changes requested):
I'd prefer to not use the workaround on the `websockets` implementation 🤔

There's already an issue on `websockets` about this, and we already documented the limitation. Let's modify the note to specify that it does not work on `websockets` implementation and add the headers to `wsproto`.

iudeen commented:
@Kludex, agreed will make the changes today! thanks for your feedback

Kludex commented:
Thanks 🙏

iudeen committed:
Merge branch 'encode:master' into master

Commit:
revert(websockets): removed work-around logic to add headers

Commit:
Revert "feat(websockets-wsproto): added server header support for websockets"

This reverts commit 62c6c7e4

Commit:
feat(wsproto): Add headers support for wsproto implementation

iudeen reviewed (commented):
implemented request changes

iudeen commented:
## Test Results:

### Code:
```python
# app.py

from fastapi import FastAPI
from starlette.websockets import WebSocket

app = FastAPI()


@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

### `Websockets` implementation without `--no-server-header`

```shell
>uvicorn app:app
INFO:     Started server process [9848]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: bxNWrBsDyxZs3+p8aaNcwTzh6ac=
Sec-WebSocket-Extensions: permessage-deflate
date: Fri, 12 Aug 2022 09:26:02 GMT
server: uvicorn
```

### `Websockets` implementation with `--no-server-header`

```shell
>uvicorn app:app --no-server-header
INFO:     Started server process [9848]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

```
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: voUjj70rAcpXQJnB1yI+0lLZuGg=
Sec-WebSocket-Extensions: permessage-deflate
date: Fri, 12 Aug 2022 09:29:49 GMT
Server: Python/3.10 websockets/10.3
```

### `Wsproto` implementation with `--no-server-header`

```shell
>uvicorn app:app --no-server-header --ws wsproto
INFO:     Started server process [9848]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

```
HTTP/1.1 101
Upgrade: WebSocket
Connection: Upgrade
Sec-WebSocket-Accept: u1f5bQpPsn8/VPZAo5ynbrRRkag=
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits=15
date: Fri, 12 Aug 2022 17:33:05 GMT
```

### `Wsproto` implementation without `--no-server-header`

```shell
>uvicorn app:app  --ws wsproto
INFO:     Started server process [9848]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

```
HTTP/1.1 101
Upgrade: WebSocket
Connection: Upgrade
Sec-WebSocket-Accept: r0+014Ywn5SbmlY9tNSEPg6FRfg=
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits=15
date: Fri, 12 Aug 2022 17:34:04 GMT
server: uvicorn
```

Commit:
docs: added limitation about Websockets - no-date-header

--no-date-header is also not compatible with Websockets

Commit:
docs: added limitation about Websockets - no-date-header

--no-date-header is also not compatible with Websockets

iudeen commented:
`--no-date-header` is not compatible with Websockets implementation. Added that also as a limitation in documentation.

Kludex commented on uvicorn/protocols/websockets/wsproto_impl.py:
```diff
@@ -248,7 +249,15 @@ async def send(self, message):
+                        continue
+                    _added_names.append(name.lower())
+                    extra_headers.append((name, value))
+
```
Why not just this? 🤔 
```suggestion
                extra_headers = list(message.get("headers", [])) + self.default_headers
```

iudeen replied:
if the user passes custom Server header, it lead to an exception of duplicate header names. That's why added a deduplication logic

Kludex replied:
What exception?

iudeen replied:
In `websockets` implementation 

```python
  File "...\lib\site-packages\websockets\datastructures.py", line 119, in __getitem__
    raise MultipleValuesError(key)
websockets.datastructures.MultipleValuesError: 'Server'
```

In `wsproto` we get duplicate `Server` headers. 

This happens when you add 

```
@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept(headers=[(b"Server", b"this is my server")])
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

If its okay to have multiple headers when using `wsproto`, I can remove the logic.

Kludex replied:
This should be a bug on `websockets` side. 🤔

It's ok to have multiple headers with the same key, it's the behavior we have on `http` protocols.

iudeen replied:
Oh, okay! will do the necessary changes. And we should add that custom `Server` header is not allowed for `websockets`.  Or should I keep the validation for Websockets for now?

Kludex replied:
Let's not allow on `websockets`.

Kludex commented:
We also need tests here, jfyk.

iudeen committed:
Update uvicorn/protocols/websockets/wsproto_impl.py

Suggestion removes duplication check on header names to keep it consistent with http implementations.

Co-authored-by: Marcelo Trylesinski <marcelotryle@gmail.com>

Commit:
revert: removed default header logic

Commit:
test: added test for wsproto implementation

Commit:
style: fixed linting

Commit:
test: added test for multiple server headers

iudeen committed:
Feature/new (#2)

Co-authored-by: Marcelo Trylesinski <marcelotryle@gmail.com>
Co-authored-by: Irfanuddin <irfanuddinshafi@gmail.com>

iudeen committed:
flake-8 corrections

iudeen committed:
run black

iudeen commented:
Closing this request, as it was made from `master` branch of the forked repo. Sorry!

Please refer to https://github.com/encode/uvicorn/pull/1606 for further updates.
